### ISSUE TITLE
Issue with firmwares 5.50 and older

### ISSUE THREAD

**Issue by Kane610** [Kane610/axis#55]
Originally reported in HASS repo https://github.com/home-assistant/core/issues/42415

There is an issue with the transition to httpx and firmwares of 5.50 and older. It works perfectly fine on newer firmwares.

Tested on a 5.75 which have Apache webserver rather than Boa and there are more differences when looking at network traces. Looking at a web request with unauthorised response setting up digest there is much more data related to the same session than on a device with boa, with boa every session ends after web server response, so looking in wire shark gives a much narrower scope per following.

Requests seemed to handle all this perfectly fine down to 5.20 or even earlier. So I think httpx is sensitive to what responses is sent from the Boa webserver.

The change from requests to httpx: https://github.com/Kane610/axis/pull/52/files
The change to async httpx: https://github.com/Kane610/axis/pull/53/files

**Comment by Kane610** [comment]
Session is created https://github.com/Kane610/axis/blob/529e43440c4c577e5588a8d6c57506f5e20472f6/axis/vapix.py#L40

Doing this call https://github.com/Kane610/axis/blob/529e43440c4c577e5588a8d6c57506f5e20472f6/axis/vapix.py#L203
yield a raised transport error from https://github.com/Kane610/axis/blob/529e43440c4c577e5588a8d6c57506f5e20472f6/axis/vapix.py#L213

**Comment by Kane610** [comment]
Succeeding using requests

Session is created https://github.com/Kane610/axis/blob/e3bfe729132ca29a44865156a361a00c0ce28a5f/axis/configuration.py#L23

Retrieving users https://github.com/Kane610/axis/blob/e3bfe729132ca29a44865156a361a00c0ce28a5f/axis/vapix.py#L184 does not yield a transport error here https://github.com/Kane610/axis/blob/e3bfe729132ca29a44865156a361a00c0ce28a5f/axis/utils.py#L15

**Comment by Kane610** [comment]
I was looking at moving from using get to post when retrieving users and noticed that this also failed, in this case both on 5.51 and 5.75. Going to even newer seems to work though.

**Comment by florimondmanca** [comment]
@Kane610 I'd recommend starting from a minimal test script and then ramping up to the actual code in the codebase. For example, what does a simple GET request to the device look like, does it succeed? If so, build a client like in the actual codebase, retry, and slowly make your way from there... Until you find at what point exactly things start looking off. Then share what this "minimal reproducible example" looks like, stripping away anything that doesn't make the bug go away.

If you haven't already you can review our Requests Compatibility guide in case that could give you som hints: https://www.python-httpx.org/compatibility/

**Comment by Kane610** [comment]
@florimondmanca absolutely i will try to reproduce it using a simple script. Indeed I followed the requests pages you'd written. And as stated above it works on newer devices so there is probably som quirk the older firmwares do that requests handle but httpx doesn't.

**Comment by Kane610** [comment]
@florimondmanca I've created the smallest thing to reproduce this issue, it fails as expected on some older devices and works fine on newer.

```
import httpx
client = httpx.Client(auth=httpx.DigestAuth("root", "pass"))
res = client.request("get", "http://10.0.0.254:80/axis-cgi/pwdgrp.cgi?action=get")
res.raise_for_status()
```

**Comment by yuriy1337** [comment]
I'm getting the same error with:
```
AXIS M3004 Network Camera
Firmware version: 5.51.7.2
```

**Comment by cdeler** [comment]
Hello @Kane610 

I checked the PR to `h11` with the real device. I mentioned the results [there](https://github.com/python-hyper/h11/pull/115#issuecomment-736699892):

#### requests lib:
```
In [1]: import requests
In [2]: from requests.auth import HTTPDigestAuth
In [3]: url = "http://[REDACTED]/axis-cgi/pwdgrp.cgi?action=get"
In [4]: requests.get(url, auth=HTTPDigestAuth("[REDACTED]", "[REDACTED]"))
Out[4]: <Response [200]>
```

#### without fix:
```
In [1]: import httpx
In [2]: url = "http://[REDACTED]/axis-cgi/pwdgrp.cgi?action=get"
In [3]: auth = httpx.DigestAuth("[REDACTED]", "[REDACTED]")
In [4]: client = httpx.Client(auth=auth)
In [5]: client.get(url)
RemoteProtocolError: peer unexpectedly closed connection
```

```
#### with fix:
In [1]: import httpx
In [2]: url = "http://[REDACTED]/axis-cgi/pwdgrp.cgi?action=get"
In [3]: auth = httpx.DigestAuth("[REDACTED]", "[REDACTED]")
In [4]: client = httpx.Client(auth=auth)
In [5]: client.get(url)
Out[5]: <Response [200 OK]>
```

**Comment by Kane610** [comment]
Awesome! Thanks for the update @cdeler

**Comment by Kane610** [comment]
I've verified the fix to work on both a 5.51 device and a 5.75 and additionally that newer firmwares keep on working 🎉

**Comment by Kane610** [comment]
This has been solved upstream so there is no need to keep this issue open anymore.
